File: ios/Sources/EmojiPicker/Presenter/WebFallbackEmojiPickerPresenter.swift
import UIKit

/// Presents the web bottom sheet inside the app's own Capacitor webview by evaluating JS there
/// (registered by `registerNativeWebBridge()` in the JS layer), and resolves once that JS
/// reports back via `handleBridgeMessage`. Kept independent of any real `WKWebView` type so it
/// stays unit-testable: production wiring in `EmojiPicker.load()` supplies a JS evaluator backed
/// by `WKWebView.evaluateJavaScript`, tests supply a fake one.
final class WebFallbackEmojiPickerPresenter: EmojiPickerPresenter {
    /// How long to wait for the JS side to report back before giving up.
    static let timeoutSeconds: TimeInterval = 3

    /// Evaluates JS in the app's webview. The completion callback fires once the script has at
    /// least started executing (i.e. the bridge/webview is alive) - it is a liveness signal, not
    /// a signal that the picker has settled.
    private let jsEvaluator: (String, @escaping () -> Void) -> Void
    private let scheduler: (TimeInterval, @escaping () -> Void) -> Void
    private var pending: [String: PendingRequest] = [:]

    convenience init(jsEvaluator: @escaping (String, @escaping () -> Void) -> Void) {
        self.init(jsEvaluator: jsEvaluator) { delay, work in
            DispatchQueue.main.asyncAfter(deadline: .now() + delay, execute: work)
        }
    }

    /// Internal: lets tests inject a fake scheduler instead of waiting real time.
    init(
        jsEvaluator: @escaping (String, @escaping () -> Void) -> Void,
        scheduler: @escaping (TimeInterval, @escaping () -> Void) -> Void
    ) {
        self.jsEvaluator = jsEvaluator
        self.scheduler = scheduler
    }

    func present(options: EmojiPickerPresentOptions, completion: @escaping (Result<EmojiPickerResult, EmojiPickerError>) -> Void) {
        // Capacitor plugin calls arrive on a background "bridge" queue, but `jsEvaluator` ends up
        // calling `WKWebView.evaluateJavaScript`, a main-thread-only UI API (mirrors the
        // unconditional hop `NativeEmojiPickerPresenter` already does for its own UIKit calls).
        DispatchQueue.main.async { [weak self] in
            self?.presentOnMainThread(options: options, completion: completion)
        }
    }

    private func presentOnMainThread(options: EmojiPickerPresentOptions, completion: @escaping (Result<EmojiPickerResult, EmojiPickerError>) -> Void) {
        let requestId = UUID().uuidString

        // Genuine timeout - i.e. the eval-completion callback below never fired at all (extremely
        // rare: webview torn down mid-call, etc). Defensively tell the JS side to close (in case
        // the bridge partially works) and settle via the same path as a normal dismiss. If the
        // eval completion DID fire by the time this runs, it's a no-op: the request just stays
        // pending for its real result.
        scheduler(Self.timeoutSeconds) { [weak self] in
            self?.handleTimeout(requestId: requestId)
        }

        // Belt-and-suspenders, mirroring `NativeEmojiPickerPresenter`'s own backgrounding safety
        // net: force-dismiss (and tell the JS sheet to close) if the app backgrounds mid-presentation,
        // since there's no guarantee the JS side's own bridge call ever lands in that case.
        let resignActiveObserver = NotificationCenter.default.addObserver(
            forName: UIApplication.willResignActiveNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            self?.dismiss(requestId: requestId)
        }

        pending[requestId] = PendingRequest(completion: completion, resignActiveObserver: resignActiveObserver)

        let json = Self.encodeOptionsJson(
            dismissOnBackdropTap: options.dismissOnBackdropTap,
            closeButton: options.closeButton,
            backdrop: options.backdrop,
            theme: options.theme
        )
        // The eval "completion" here only means the script at least started executing (i.e. the
        // bridge/webview is alive) - it cancels the timeout WITHOUT settling the request, which
        // stays pending for the real result reported later via handleBridgeMessage.
        jsEvaluator("window.__CapacitorEmojiPickerPresentWeb('\(requestId)', '\(json)')") { [weak self] in
            self?.pending[requestId]?.evalAcknowledged = true
        }
    }

    /// Called by the `WKScriptMessageHandler` once the JS sheet settles.
    func handleBridgeMessage(requestId: String, emoji: String?, errorCode: String?) {
        settle(requestId: requestId, emoji: emoji, errorCode: errorCode)
    }

    /// Force-dismisses a still-pending presentation (e.g. on app backgrounding).
    func dismiss(requestId: String) {
        jsEvaluator("window.__CapacitorEmojiPickerDismissWeb && window.__CapacitorEmojiPickerDismissWeb('\(requestId)')") {}
        settle(requestId: requestId, emoji: nil, errorCode: nil)
    }

    /// Handles the timeout firing: a no-op if the eval-completion callback already acknowledged
    /// the bridge is alive (the request stays pending for its real result); otherwise this is a
    /// genuine bridge-liveness failure (the JS side never even started processing the request),
    /// so defensively evaluate the JS dismiss call and settle as an error rather than reusing
    /// `dismiss(requestId:)`'s success/null semantics (that's reserved for genuine user/app
    /// dismissal, e.g. backgrounding).
    private func handleTimeout(requestId: String) {
        guard let request = pending[requestId], !request.evalAcknowledged else { return }
        jsEvaluator("window.__CapacitorEmojiPickerDismissWeb && window.__CapacitorEmojiPickerDismissWeb('\(requestId)')") {}
        settle(requestId: requestId, emoji: nil, errorCode: ErrorCodes.notImplemented)
    }

    private func settle(requestId: String, emoji: String?, errorCode: String?) {
        guard let request = pending.removeValue(forKey: requestId) else { return }
        NotificationCenter.default.removeObserver(request.resignActiveObserver)
        if let errorCode = errorCode {
            request.completion(.failure(EmojiPickerError(code: errorCode)))
        } else {
            request.completion(.success(EmojiPickerResult(emoji: emoji)))
        }
    }

    /// Hand-rolled instead of `JSONSerialization`: `size`/`position`/`theme` are always one of a
    /// small fixed set of ASCII enum values, and `backdrop.color` is a hex string matched against
    /// a strict pattern, all validated/defaulted when the plugin call is parsed, never arbitrary
    /// user text, so plain string interpolation is safe here.
    private static func encodeOptionsJson(
        dismissOnBackdropTap: Bool,
        closeButton: EmojiCloseButtonOptions,
        backdrop: EmojiBackdropOptions,
        theme: String
    ) -> String {
        "{\"dismissOnBackdropTap\":\(dismissOnBackdropTap),"
            + "\"closeButton\":{\"size\":\"\(closeButton.size)\",\"position\":\"\(closeButton.position)\",\"hidden\":\(closeButton.hidden)},"
            + "\"backdrop\":{\"color\":\"\(backdrop.color)\",\"blur\":\(backdrop.blur)},"
            + "\"theme\":\"\(theme)\"}"
    }

    private struct PendingRequest {
        let completion: (Result<EmojiPickerResult, EmojiPickerError>) -> Void
        let resignActiveObserver: NSObjectProtocol
        var evalAcknowledged = false
    }
}

File: android/src/main/java/app/independo/capacitoremojipicker/presenter/WebFallbackEmojiPickerPresenter.java
package app.independo.capacitoremojipicker.presenter;

import android.app.Activity;
import android.os.Handler;
import android.os.Looper;
import androidx.lifecycle.DefaultLifecycleObserver;
import androidx.lifecycle.LifecycleOwner;
import app.independo.capacitoremojipicker.core.EmojiBackdropOptions;
import app.independo.capacitoremojipicker.core.EmojiCloseButtonOptions;
import app.independo.capacitoremojipicker.core.EmojiPickerCallback;
import app.independo.capacitoremojipicker.core.EmojiPickerResult;
import app.independo.capacitoremojipicker.core.ErrorCodes;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

/**
 * Presents the web bottom sheet inside the app's own Capacitor webview by evaluating JS there
 * (registered by {@code registerNativeWebBridge()} in the JS layer), and resolves once that JS
 * reports back via {@link #onWebResult}. Kept independent of any real {@code Bridge}/{@code
 * WebView} type so it stays unit-testable without the Android Gradle Plugin's mockable
 * {@code android.jar}: production wiring in {@code EmojiPicker#load()} supplies a {@link
 * JsEvaluator} backed by {@code Bridge#eval} and an {@link ActivityProvider}; tests supply fakes.
 */
public class WebFallbackEmojiPickerPresenter implements EmojiPickerPresenter {

    /**
     * Evaluates JS in the app's webview. Abstracts away the real Capacitor {@code Bridge}.
     * {@code onEvaluated} fires once the script has at least started executing (i.e. the
     * bridge/webview is alive) - it is a liveness signal, not a signal that the picker has
     * settled.
     */
    public interface JsEvaluator {
        void eval(String js, Runnable onEvaluated);
    }

    /** Supplies the current hosting Activity; returns null if none is available. */
    public interface ActivityProvider {
        Activity get();
    }

    /** Posts/cancels delayed work. Abstracts away {@code Handler} for tests. */
    interface Scheduler {
        void postDelayed(Runnable runnable, long delayMillis);
        void cancel(Runnable runnable);
    }

    /**
     * Hops onto the UI thread to run {@code work}. Abstracts away {@code Activity#runOnUiThread}
     * (a {@code final} method, so it can't be overridden by a test fake {@code Activity} subclass)
     * for tests.
     */
    interface UiThreadDispatcher {
        void runOnUiThread(Activity activity, Runnable work);
    }

    /** How long to wait for the JS side to report back before giving up. */
    static final long TIMEOUT_MILLIS = 3000;

    private final JsEvaluator jsEvaluator;
    private final ActivityProvider activityProvider;
    private final Scheduler scheduler;
    private final UiThreadDispatcher uiThreadDispatcher;
    private final Map<String, PendingRequest> pending = new HashMap<>();

    public WebFallbackEmojiPickerPresenter(JsEvaluator jsEvaluator, ActivityProvider activityProvider) {
        this(jsEvaluator, activityProvider, defaultScheduler(), defaultUiThreadDispatcher());
    }

    /** Package-private: lets tests inject a fake {@link Scheduler} and {@link UiThreadDispatcher}. */
    WebFallbackEmojiPickerPresenter(
        JsEvaluator jsEvaluator,
        ActivityProvider activityProvider,
        Scheduler scheduler,
        UiThreadDispatcher uiThreadDispatcher
    ) {
        this.jsEvaluator = jsEvaluator;
        this.activityProvider = activityProvider;
        this.scheduler = scheduler;
        this.uiThreadDispatcher = uiThreadDispatcher;
    }

    private static Scheduler defaultScheduler() {
        Handler handler = new Handler(Looper.getMainLooper());
        return new Scheduler() {
            @Override
            public void postDelayed(Runnable runnable, long delayMillis) {
                handler.postDelayed(runnable, delayMillis);
            }

            @Override
            public void cancel(Runnable runnable) {
                handler.removeCallbacks(runnable);
            }
        };
    }

    // Capacitor invokes @PluginMethod handlers on a background "CapacitorPlugins" thread, but
    // Lifecycle#addObserver() is a UI-thread-only API (mirrors the same hop
    // NativeEmojiPickerPresenter already does for its own UI-thread-only calls). Hop onto the UI
    // thread whenever we have a real Activity to hop through; a null Activity means there's no UI
    // to touch, so run synchronously (no Lifecycle to register against either way).
    private static UiThreadDispatcher defaultUiThreadDispatcher() {
        return (activity, work) -> {
            if (activity != null) {
                activity.runOnUiThread(work);
            } else {
                work.run();
            }
        };
    }

    @Override
    public void present(
        String presentation,
        boolean dismissOnBackdropTap,
        EmojiCloseButtonOptions closeButton,
        EmojiBackdropOptions backdrop,
        String theme,
        EmojiPickerCallback callback
    ) {
        Activity activity = activityProvider.get();
        uiThreadDispatcher.runOnUiThread(
            activity,
            () -> presentOnUiThread(activity, dismissOnBackdropTap, closeButton, backdrop, theme, callback)
        );
    }

    private void presentOnUiThread(
        Activity activity,
        boolean dismissOnBackdropTap,
        EmojiCloseButtonOptions closeButton,
        EmojiBackdropOptions backdrop,
        String theme,
        EmojiPickerCallback callback
    ) {
        String requestId = UUID.randomUUID().toString();

        // Genuine timeout - i.e. the eval-completion callback below never fired at all (extremely
        // rare: webview torn down mid-call, etc). This means the JS side never even started
        // processing the request, so it's a bridge-liveness failure, not a user dismissal:
        // defensively tell the JS side to close (in case the bridge partially works), but settle
        // as an error rather than reusing dismiss()'s success/null semantics.
        Runnable timeoutRunnable = () -> {
            jsEvaluator.eval("window.__CapacitorEmojiPickerDismissWeb && window.__CapacitorEmojiPickerDismissWeb('" + requestId + "')", () -> {});
            settle(requestId, null, ErrorCodes.NOT_IMPLEMENTED);
        };

        DefaultLifecycleObserver lifecycleObserver = null;
        LifecycleOwner lifecycleOwner = null;
        if (activity instanceof LifecycleOwner) {
            lifecycleOwner = (LifecycleOwner) activity;
            lifecycleObserver = new DefaultLifecycleObserver() {
                @Override
                public void onPause(LifecycleOwner owner) {
                    dismiss(requestId);
                }

                @Override
                public void onDestroy(LifecycleOwner owner) {
                    dismiss(requestId);
                }
            };
            lifecycleOwner.getLifecycle().addObserver(lifecycleObserver);
        }

        pending.put(requestId, new PendingRequest(callback, timeoutRunnable, lifecycleObserver, lifecycleOwner));
        scheduler.postDelayed(timeoutRunnable, TIMEOUT_MILLIS);

        // The eval "completion" here only means the script at least started executing (i.e. the
        // bridge/webview is alive) - it cancels the timeout WITHOUT settling the request, which
        // stays pending for the real result reported later via onWebResult.
        jsEvaluator.eval(
            "window.__CapacitorEmojiPickerPresentWeb('" + requestId + "', '"
                + encodeOptionsBase64Free(dismissOnBackdropTap, closeButton, backdrop, theme) + "')",
            () -> scheduler.cancel(timeoutRunnable)
        );
    }

    /** Called by the JS-interface bridge object once the JS sheet settles. */
    public void onWebResult(String requestId, String emoji, String errorCode) {
        settle(requestId, emoji, errorCode);
    }

    /** Force-dismisses a still-pending presentation (e.g. on Activity destruction/backgrounding). */
    public void dismiss(String requestId) {
        jsEvaluator.eval("window.__CapacitorEmojiPickerDismissWeb && window.__CapacitorEmojiPickerDismissWeb('" + requestId + "')", () -> {});
        settle(requestId, null, null);
    }

    private void settle(String requestId, String emoji, String errorCode) {
        PendingRequest request = pending.remove(requestId);
        if (request == null) {
            return;
        }
        scheduler.cancel(request.timeoutRunnable);
        if (request.lifecycleObserver != null && request.lifecycleOwner != null) {
            request.lifecycleOwner.getLifecycle().removeObserver(request.lifecycleObserver);
        }
        if (errorCode != null) {
            request.callback.onError(errorCode);
        } else {
            request.callback.onResult(new EmojiPickerResult(emoji));
        }
    }

    /**
     * Hand-rolled instead of a JSON library: `size`/`position`/`theme` are always one of a small
     * fixed set of ASCII enum values, and `backdrop.color` is a hex string matched against a
     * strict pattern, all validated/defaulted in {@code EmojiPicker#present}, never arbitrary
     * user text, so plain string interpolation is safe here. This also sidesteps
     * `org.json`/`android.util.Base64` throwing under the mockable `android.jar` in plain JUnit
     * unit tests (see the plan's Global Constraints).
     */
    private static String encodeOptionsBase64Free(
        boolean dismissOnBackdropTap,
        EmojiCloseButtonOptions closeButton,
        EmojiBackdropOptions backdrop,
        String theme
    ) {
        return "{"
            + "\"dismissOnBackdropTap\":" + dismissOnBackdropTap + ","
            + "\"closeButton\":{"
            + "\"size\":\"" + closeButton.size + "\","
            + "\"position\":\"" + closeButton.position + "\","
            + "\"hidden\":" + closeButton.hidden
            + "},"
            + "\"backdrop\":{"
            + "\"color\":\"" + backdrop.color + "\","
            + "\"blur\":" + backdrop.blur
            + "},"
            + "\"theme\":\"" + theme + "\""
            + "}";
    }

    private static final class PendingRequest {
        final EmojiPickerCallback callback;
        final Runnable timeoutRunnable;
        final DefaultLifecycleObserver lifecycleObserver;
        final LifecycleOwner lifecycleOwner;

        PendingRequest(
            EmojiPickerCallback callback,
            Runnable timeoutRunnable,
            DefaultLifecycleObserver lifecycleObserver,
            LifecycleOwner lifecycleOwner
        ) {
            this.callback = callback;
            this.timeoutRunnable = timeoutRunnable;
            this.lifecycleObserver = lifecycleObserver;
            this.lifecycleOwner = lifecycleOwner;
        }
    }
}
